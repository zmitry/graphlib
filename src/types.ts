declare module "graph" {
  export interface GraphOptions {
    directed?: boolean; // default: true.
    multigraph?: boolean; // default: false.
    compound?: boolean; // default: false.
  }

  export interface Edge {
    v: string;
    w: string;
    /** The name that uniquely identifies a multi-edge. */
    name?: string;
  }

  export class Graph<NodeLabel = any, EdgeLabel = any> {
    constructor(options?: GraphOptions);

    /**
     * Sets the default node label. This label will be assigned as default label
     * in case if no label was specified while setting a node.
     * Complexity: O(1).
     *
     * @argument label - default node label.
     * @returns the graph, allowing this to be chained with other functions.
     */
    setDefaultNodeLabel(label: NodeLabel): Graph<NodeLabel, EdgeLabel>;

    /**
     * Sets the default node label factory function. This function will be invoked
     * each time when setting a node with no label specified and returned value
     * will be used as a label for node.
     * Complexity: O(1).
     *
     * @argument labelFn - default node label factory function.
     * @returns the graph, allowing this to be chained with other functions.
     */
    setDefaultNodeLabel(
      labelFn: (v: string) => NodeLabel
    ): Graph<NodeLabel, EdgeLabel>;

    /**
     * Creates or updates the value for the node v in the graph. If label is supplied
     * it is set as the value for the node. If label is not supplied and the node was
     * created by this call then the default node label will be assigned.
     * Complexity: O(1).
     *
     * @argument name - node name.
     * @argument label - value to set for node.
     * @returns the graph, allowing this to be chained with other functions.
     */
    setNode(name: string, label?: NodeLabel): Graph<NodeLabel, EdgeLabel>;

    /**
     * Invokes setNode method for each node in names list.
     * Complexity: O(|names|).
     *
     * @argument names - list of nodes names to be set.
     * @argument label - value to set for each node in list.
     * @returns the graph, allowing this to be chained with other functions.
     */
    setNodes(names: string[], label?: NodeLabel): Graph<NodeLabel, EdgeLabel>;

    /**
     * Sets node p as a parent for node v if it is defined, or removes the
     * parent for v if p is undefined. Method throws an exception in case of
     * invoking it in context of noncompound graph.
     * Average-case complexity: O(1).
     *
     * @argument v - node to be child for p.
     * @argument p - node to be parent for v.
     * @returns the graph, allowing this to be chained with other functions.
     */
    setParent(v: string, p?: string): Graph<NodeLabel, EdgeLabel>;

    /**
     * Gets parent node for node v.
     * Complexity: O(1).
     *
     * @argument v - node to get parent of.
     * @returns parent node name or void if v has no parent.
     */
    parent(v: string): string | void;

    /**
     * Gets list of direct children of node v.
     * Complexity: O(1).
     *
     * @argument v - node to get children of.
     * @returns children nodes names list.
     */
    children(v: string): string[];

    /**
     * Creates new graph with nodes filtered via filter. Edges incident to rejected node
     * are also removed. In case of compound graph, if parent is rejected by filter,
     * than all its children are rejected too.
     * Average-case complexity: O(|E|+|V|).
     *
     * @argument filter - filtration function detecting whether the node should stay or not.
     * @returns new graph made from current and nodes filtered.
     */
    filterNodes(filter: (v: string) => boolean): Graph<NodeLabel, EdgeLabel>;

    /**
     * Sets the default edge label. This label will be assigned as default label
     * in case if no label was specified while setting an edge.
     * Complexity: O(1).
     *
     * @argument label - default edge label.
     * @returns the graph, allowing this to be chained with other functions.
     */
    setDefaultEdgeLabel(label: EdgeLabel): Graph<NodeLabel, EdgeLabel>;

    /**
     * Sets the default edge label factory function. This function will be invoked
     * each time when setting an edge with no label specified and returned value
     * will be used as a label for edge.
     * Complexity: O(1).
     *
     * @argument labelFn - default edge label factory function.
     * @returns the graph, allowing this to be chained with other functions.
     */
    setDefaultEdgeLabel(
      labelFn: (v: string) => EdgeLabel
    ): Graph<NodeLabel, EdgeLabel>;

    /**
     * Establish an edges path over the nodes in nodes list. If some edge is already
     * exists, it will update its label, otherwise it will create an edge between pair
     * of nodes with label provided or default label if no label provided.
     * Complexity: O(|nodes|).
     *
     * @argument nodes - list of nodes to be connected in series.
     * @argument label - value to set for each edge between pairs of nodes.
     * @returns the graph, allowing this to be chained with other functions.
     */
    setPath(nodes: string[], label?: EdgeLabel): Graph<NodeLabel, EdgeLabel>;

    /**
     * Detects whether graph has a node with specified name or not.
     *
     * @argument name - name of the node.
     * @returns true if graph has node with specified name, false - otherwise.
     */
    hasNode(name: string): boolean;

    /**
     * Remove the node with the name from the graph or do nothing if the node is not in
     * the graph. If the node was removed this function also removes any incident
     * edges.
     * Complexity: O(1).
     *
     * @argument name - name of the node.
     * @returns the graph, allowing this to be chained with other functions.
     */
    removeNode(name: string): Graph<NodeLabel, EdgeLabel>;

    /**
     * Gets all nodes of the graph. Note, the in case of compound graph subnodes are
     * not included in list.
     * Complexity: O(1).
     *
     * @returns list of graph nodes.
     */
    nodes(): string[];

    /**
     * Gets the label of node with specified name.
     * Complexity: O(|V|).
     *
     * @returns label value of the node.
     */
    node(name: string): NodeLabel;

    /**
     * Creates or updates the label for the edge (v, w) with the optionally supplied
     * name. If label is supplied it is set as the value for the edge. If label is not
     * supplied and the edge was created by this call then the default edge label will
     * be assigned. The name parameter is only useful with multigraphs.
     * Complexity: O(1).
     *
     * @argument v - edge source node.
     * @argument w - edge sink node.
     * @argument label - value to associate with the edge.
     * @argument name - unique name of the edge in order to identify it in multigraph.
     * @returns the graph, allowing this to be chained with other functions.
     */
    setEdge(
      v: string,
      w: string,
      label?: EdgeLabel,
      name?: string
    ): Graph<NodeLabel, EdgeLabel>;

    /**
     * Creates or updates the label for the specified edge. If label is supplied it is
     * set as the value for the edge. If label is not supplied and the edge was created
     * by this call then the default edge label will be assigned. The name parameter is
     * only useful with multigraphs.
     * Complexity: O(1).
     *
     * @argument edge - edge descriptor.
     * @argument label - value to associate with the edge.
     * @returns the graph, allowing this to be chained with other functions.
     */
    setEdge(edge: Edge, label?: EdgeLabel): Graph<NodeLabel, EdgeLabel>;

    /**
     * Gets edges of the graph. In case of compound graph subgraphs are not considered.
     * Complexity: O(|E|).
     *
     * @return graph edges list.
     */
    edges(): Edge[];

    /**
     * Gets the label for the specified edge.
     * Complexity: O(1).
     *
     * @argument v - edge source node.
     * @argument w - edge sink node.
     * @argument name - name of the edge (actual for multigraph<any,any>).
     * @returns value associated with specified edge.
     */
    edge(v: string, w: string, name?: string): EdgeLabel;

    /**
     * Gets the label for the specified edge.
     * Complexity: O(1).
     *
     * @argument edge - edge descriptor.
     * @returns value associated with specified edge.
     */
    edge(e: Edge): EdgeLabel;

    /**
     * Detects whether the graph contains specified edge or not. No subgraphs are considered.
     * Complexity: O(1).
     *
     * @argument v - edge source node.
     * @argument w - edge sink node.
     * @argument name - name of the edge (actual for multigraph<any,any>).
     * @returns whether the graph contains the specified edge or not.
     */
    hasEdge(v: string, w: string, name?: string): boolean;

    /**
     * Detects whether the graph contains specified edge or not. No subgraphs are considered.
     * Complexity: O(1).
     *
     * @argument edge - edge descriptor.
     * @returns whether the graph contains the specified edge or not.
     */
    hasEdge(edge: Edge): boolean;

    /**
     * Removes the specified edge from the graph. No subgraphs are considered.
     * Complexity: O(1).
     *
     * @argument edge - edge descriptor.
     * @returns the graph, allowing this to be chained with other functions.
     */
    removeEdge(edge: Edge): Graph<NodeLabel, EdgeLabel>;

    /**
     * Removes the specified edge from the graph. No subgraphs are considered.
     * Complexity: O(1).
     *
     * @argument v - edge source node.
     * @argument w - edge sink node.
     * @argument name - name of the edge (actual for multigraph<any,any>).
     * @returns the graph, allowing this to be chained with other functions.
     */
    removeEdge(
      v: string,
      w: string,
      name?: string
    ): Graph<NodeLabel, EdgeLabel>;

    /**
     * Return all edges that point to the node v. Optionally filters those edges down to just those
     * coming from node u. Behavior is undefined for undirected graphs - use nodeEdges instead.
     * Complexity: O(|E|).
     *
     * @argument v - edge sink node.
     * @argument w - edge source node.
     * @returns edges descriptors list if v is in the graph, or undefined otherwise.
     */
    inEdges(v: string, w?: string): void | Edge[];

    /**
     * Return all edges that are pointed at by node v. Optionally filters those edges down to just
     * those point to w. Behavior is undefined for undirected graphs - use nodeEdges instead.
     * Complexity: O(|E|).
     *
     * @argument v - edge source node.
     * @argument w - edge sink node.
     * @returns edges descriptors list if v is in the graph, or undefined otherwise.
     */
    outEdges(v: string, w?: string): void | Edge[];

    /**
     * Returns all edges to or from node v regardless of direction. Optionally filters those edges
     * down to just those between nodes v and w regardless of direction.
     * Complexity: O(|E|).
     *
     * @argument v - edge adjacent node.
     * @argument w - edge adjacent node.
     * @returns edges descriptors list if v is in the graph, or undefined otherwise.
     */
    nodeEdges(v: string, w?: string): void | Edge[];

    /**
     * Return all nodes that are predecessors of the specified node or undefined if node v is not in
     * the graph. Behavior is undefined for undirected graphs - use neighbors instead.
     * Complexity: O(|V|).
     *
     * @argument v - node identifier.
     * @returns node identifiers list or undefined if v is not in the graph.
     */
    predecessors(v: string): void | string[];

    /**
     * Return all nodes that are successors of the specified node or undefined if node v is not in
     * the graph. Behavior is undefined for undirected graphs - use neighbors instead.
     * Complexity: O(|V|).
     *
     * @argument v - node identifier.
     * @returns node identifiers list or undefined if v is not in the graph.
     */
    successors(v: string): void | string[];

    /**
     * Return all nodes that are predecessors or successors of the specified node or undefined if
     * node v is not in the graph.
     * Complexity: O(|V|).
     *
     * @argument v - node identifier.
     * @returns node identifiers list or undefined if v is not in the graph.
     */

    neighbors(v: string): void | string[];

    /**
     * Whether graph was created with 'directed' flag set to true or not.
     *
     * @returns whether the graph edges have an orientation.
     */
    isDirected(): boolean;

    /**
     * Whether graph was created with 'multigraph' flag set to true or not.
     *
     * @returns whether the pair of nodes of the graph can have multiple edges.
     */
    isMultigraph(): boolean;

    /**
     * Whether graph was created with 'compound' flag set to true or not.
     *
     * @returns whether a node of the graph can have subnodes.
     */
    isCompound(): boolean;

    /**
     * Sets the label of the graph.
     *
     * @argument label - label value.
     * @returns the graph, allowing this to be chained with other functions.
     */
    setGraph(label: string): Graph<NodeLabel, EdgeLabel>;

    /**
     * Gets the graph label.
     *
     * @returns currently assigned label for the graph or undefined if no label assigned.
     */
    graph(): void | string;

    /**
     * Gets the number of nodes in the graph.
     * Complexity: O(1).
     *
     * @returns nodes count.
     */
    nodeCount(): number;

    /**
     * Gets the number of edges in the graph.
     * Complexity: O(1).
     *
     * @returns edges count.
     */
    edgeCount(): number;

    /**
     * Gets list of nodes without in-edges.
     * Complexity: O(|V|).
     *
     * @returns the graph source nodes.
     */
    sources(): string[];

    /**
     * Gets list of nodes without out-edges.
     * Complexity: O(|V|).
     *
     * @returns the graph source nodes.
     */
    sinks(): string[];
  }
  export function writeJSON(graph: Graph): Object;
  export function readJSON<N = any, L = any>(graphJSON: object): Graph<N, L>;
  export function writeDot<N, L>(g: Graph<N, L>): string;
}
