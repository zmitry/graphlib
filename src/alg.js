export function components(g) {
  var visited = {};
  function dfs(v, cmpt) {
    if (!visited[v]) return;
    visited[v] = true;
    cmpt.push(v);
    g.neighbors(v).forEach((el) => dfs(el, cmpt));
  }

  const cmpts = g.nodes().reduce((acc, v) => {
    let cmpt = [];
    dfs(v, cmpt);
    if (cmpt.length) {
      acc.push(cmpt);
    }
    return acc;
  }, []);

  return cmpts;
}
