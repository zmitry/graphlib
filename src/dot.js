import { each, isObject, map } from "lodash";

const UNESCAPED_ID_PATTERN = /^[a-zA-Z\200-\377_][a-zA-Z\200-\377_0-9]*$/;

export function writeOne(g, intend = "") {
  const ec = g.isDirected() ? "->" : "--";
  const writer = makeWriter(intend);

  if (!g.isMultigraph()) {
    writer.write("strict ");
  }

  writer.writeLine(`${g.isDirected() ? "digraph" : "graph"} {`);
  writer.indent();

  const graphAttrs = g.graph();
  if (isObject(graphAttrs)) {
    each(graphAttrs, (v, k) => {
      writer.writeLine(`${id(k)}=${id(v)};`);
    });
  }

  writeSubgraph(g, undefined, writer);

  g.edges().forEach((edge) => {
    writeEdge(g, edge, ec, writer);
  });

  writer.unindent();
  writer.writeLine("}");

  return writer.toString();
}

function writeSubgraph(g, v, writer) {
  const children = g.isCompound() ? g.children(v) : g.nodes();
  each(children, (w) => {
    if (!g.isCompound() || !g.children(w).length) {
      writeNode(g, w, writer);
    } else {
      writer.writeLine(`subgraph ${id(w)} {`);
      writer.indent();

      if (isObject(g.node(w))) {
        map(g.node(w), (val, key) => {
          writer.writeLine(`${id(key)}=${id(val)};`);
        });
      }

      writeSubgraph(g, w, writer);
      writer.unindent();
      writer.writeLine("}");
    }
  });
}

function writeNode(g, v, writer) {
  writer.write(id(v));
  writeAttrs(g.node(v), writer);
  writer.writeLine();
}

function writeEdge(g, edge, ec, writer) {
  const v = edge.v;
  const w = edge.w;
  const attrs = g.edge(edge);

  writer.write(`${id(v)} ${ec} ${id(w)}`);
  writeAttrs(attrs, writer);
  writer.writeLine();
}

function writeAttrs(attrs, writer) {
  if (isObject(attrs)) {
    const attrStrs = map(attrs, (val, key) => `${id(key)}=${id(val)}`);
    if (attrStrs.length) {
      writer.write(` [${attrStrs.join(",")}]`);
    }
  }
}

function id(obj) {
  if (typeof obj === "number" || obj.toString().match(UNESCAPED_ID_PATTERN)) {
    return obj;
  }

  return `"${obj.toString().replace(/"/g, '\\"')}"`;
}

function makeWriter(INDENT = "") {
  let indent = "";
  let content = "";
  let shouldIndent = true;
  const write = (str) => {
    if (shouldIndent) {
      shouldIndent = false;
      content += indent;
    }
    content += str;
  };
  return {
    indent() {
      indent += INDENT;
    },
    unindent() {
      indent = indent.slice(INDENT.length);
    },
    toString() {
      return content;
    },
    writeLine(line = "") {
      write(`${line}\n`);
      shouldIndent = true;
    },
    write: write
  };
}
