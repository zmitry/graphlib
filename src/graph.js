import {
  has,
  constant,
  isFunction,
  keys,
  filter,
  isEmpty,
  each,
  isUndefined,
  union,
  reduce
} from "lodash";
const DEFAULT_EDGE_NAME = "\x00";
const GRAPH_NODE = "\x00";
const EDGE_KEY_DELIM = "\x01";

const values = Object.values;
// Implementation notes:
//
//  * Node id query functions should return string ids for the nodes
//  * Edge id query functions should return an "edgeObj", edge object, that is
//    composed of enough information to uniquely identify an edge: {v, w, name}.
//  * Internally we use an "edgeId", a stringified form of the edgeObj, to
//    reference edges. This is because we need a performant way to look these
//    edges up and, object properties, which have string keys, are the closest
//    we're going to get to a performant hashtable in JavaScript.

class Graph {
  constructor(opts) {
    this._isDirected = has(opts, "directed") ? opts.directed : true;
    this._isMultigraph = has(opts, "multigraph") ? opts.multigraph : false;
    this._isCompound = has(opts, "compound") ? opts.compound : false;

    // Label for the graph itself
    this._label = undefined;

    // Defaults to be set when creating a new node
    this._defaultNodeLabelFn = constant(undefined);

    // Defaults to be set when creating a new edge
    this._defaultEdgeLabelFn = constant(undefined);

    // v -> label
    this._nodes = {};

    if (this._isCompound) {
      // v -> parent
      this._parent = {};

      // v -> children
      this._children = {};
      this._children[GRAPH_NODE] = {};
    }

    // v -> edgeObj
    this._in = {};

    // u -> v -> Number
    this._preds = {};

    // v -> edgeObj
    this._out = {};

    // v -> w -> Number
    this._sucs = {};

    // e -> edgeObj
    this._edgeObjs = {};

    // e -> label
    this._edgeLabels = {};
  }

  /* === Graph functions ========= */

  isDirected() {
    return this._isDirected;
  }

  isMultigraph() {
    return this._isMultigraph;
  }

  isCompound() {
    return this._isCompound;
  }

  setGraph(label) {
    this._label = label;
    return this;
  }

  graph() {
    return this._label;
  }

  /* === Node functions ========== */

  setDefaultNodeLabel(newDefault) {
    if (!isFunction(newDefault)) {
      newDefault = constant(newDefault);
    }
    this._defaultNodeLabelFn = newDefault;
    return this;
  }

  nodeCount() {
    return this._nodeCount;
  }

  nodes() {
    return keys(this._nodes);
  }

  sources() {
    const self = this;
    return filter(this.nodes(), (v) => isEmpty(self._in[v]));
  }

  sinks() {
    const self = this;
    return filter(this.nodes(), (v) => isEmpty(self._out[v]));
  }

  setNodes(vs, value) {
    const args = arguments;
    const self = this;
    each(vs, (v) => {
      if (args.length > 1) {
        self.setNode(v, value);
      } else {
        self.setNode(v);
      }
    });
    return this;
  }

  setNode(v, value) {
    if (has(this._nodes, v)) {
      if (arguments.length > 1) {
        this._nodes[v] = value;
      }
      return this;
    }

    this._nodes[v] = arguments.length > 1 ? value : this._defaultNodeLabelFn(v);
    if (this._isCompound) {
      this._parent[v] = GRAPH_NODE;
      this._children[v] = {};
      this._children[GRAPH_NODE][v] = true;
    }
    this._in[v] = {};
    this._preds[v] = {};
    this._out[v] = {};
    this._sucs[v] = {};
    ++this._nodeCount;
    return this;
  }

  node(v) {
    return this._nodes[v];
  }

  hasNode(v) {
    return has(this._nodes, v);
  }

  removeNode(v) {
    const self = this;
    if (has(this._nodes, v)) {
      const removeEdge = (e) => {
        self.removeEdge(self._edgeObjs[e]);
      };
      delete this._nodes[v];
      if (this._isCompound) {
        this._removeFromParentsChildList(v);
        delete this._parent[v];
        each(this.children(v), (child) => {
          self.setParent(child);
        });
        delete this._children[v];
      }
      each(keys(this._in[v]), removeEdge);
      delete this._in[v];
      delete this._preds[v];
      each(keys(this._out[v]), removeEdge);
      delete this._out[v];
      delete this._sucs[v];
      --this._nodeCount;
    }
    return this;
  }

  setParent(v, parent) {
    if (!this._isCompound) {
      throw new Error("Cannot set parent in a non-compound graph");
    }

    if (isUndefined(parent)) {
      parent = GRAPH_NODE;
    } else {
      // Coerce parent to string
      parent += "";
      for (
        let ancestor = parent;
        !isUndefined(ancestor);
        ancestor = this.parent(ancestor)
      ) {
        if (ancestor === v) {
          throw new Error(
            `Setting ${parent} as parent of ${v} would create a cycle`
          );
        }
      }

      this.setNode(parent);
    }

    this.setNode(v);
    this._removeFromParentsChildList(v);
    this._parent[v] = parent;
    this._children[parent][v] = true;
    return this;
  }

  _removeFromParentsChildList(v) {
    delete this._children[this._parent[v]][v];
  }

  parent(v) {
    if (this._isCompound) {
      const parent = this._parent[v];
      if (parent !== GRAPH_NODE) {
        return parent;
      }
    }
  }

  children(v) {
    if (isUndefined(v)) {
      v = GRAPH_NODE;
    }

    if (this._isCompound) {
      const children = this._children[v];
      if (children) {
        return keys(children);
      }
    } else if (v === GRAPH_NODE) {
      return this.nodes();
    } else if (this.hasNode(v)) {
      return [];
    }
  }

  predecessors(v) {
    const predsV = this._preds[v];
    if (predsV) {
      return keys(predsV);
    }
  }

  successors(v) {
    const sucsV = this._sucs[v];
    if (sucsV) {
      return keys(sucsV);
    }
  }

  neighbors(v) {
    const preds = this.predecessors(v);
    if (preds) {
      return union(preds, this.successors(v));
    }
  }

  isLeaf(v) {
    let neighbors;
    if (this.isDirected()) {
      neighbors = this.successors(v);
    } else {
      neighbors = this.neighbors(v);
    }
    return neighbors.length === 0;
  }

  filterNodes(filter) {
    const copy = new this.constructor({
      directed: this._isDirected,
      multigraph: this._isMultigraph,
      compound: this._isCompound
    });

    copy.setGraph(this.graph());

    const self = this;
    each(this._nodes, (value, v) => {
      if (filter(v)) {
        copy.setNode(v, value);
      }
    });

    each(this._edgeObjs, (e) => {
      if (copy.hasNode(e.v) && copy.hasNode(e.w)) {
        copy.setEdge(e, self.edge(e));
      }
    });

    const parents = {};
    function findParent(v) {
      const parent = self.parent(v);
      if (parent === undefined || copy.hasNode(parent)) {
        parents[v] = parent;
        return parent;
      } else if (parent in parents) {
        return parents[parent];
      } else {
        return findParent(parent);
      }
    }

    if (this._isCompound) {
      each(copy.nodes(), (v) => {
        copy.setParent(v, findParent(v));
      });
    }

    return copy;
  }

  /* === Edge functions ========== */

  setDefaultEdgeLabel(newDefault) {
    if (!isFunction(newDefault)) {
      newDefault = constant(newDefault);
    }
    this._defaultEdgeLabelFn = newDefault;
    return this;
  }

  edgeCount() {
    return this._edgeCount;
  }

  edges() {
    return values(this._edgeObjs);
  }

  setPath(vs, value) {
    const self = this;
    const args = arguments;
    reduce(vs, (v, w) => {
      if (args.length > 1) {
        self.setEdge(v, w, value);
      } else {
        self.setEdge(v, w);
      }
      return w;
    });
    return this;
  }

  /*
   * setEdge(v, w, [value, [name]])
   * setEdge({ v, w, [name] }, [value])
   */
  setEdge(...args) {
    let v;
    let w;
    let name;
    let value;
    let valueSpecified = false;
    const arg0 = args[0];

    if (typeof arg0 === "object" && arg0 !== null && "v" in arg0) {
      v = arg0.v;
      w = arg0.w;
      name = arg0.name;
      if (args.length === 2) {
        value = args[1];
        valueSpecified = true;
      }
    } else {
      v = arg0;
      w = args[1];
      name = args[3];
      if (args.length > 2) {
        value = args[2];
        valueSpecified = true;
      }
    }

    v = `${v}`;
    w = `${w}`;
    if (!isUndefined(name)) {
      name = `${name}`;
    }

    const e = edgeArgsToId(this._isDirected, v, w, name);
    if (has(this._edgeLabels, e)) {
      if (valueSpecified) {
        this._edgeLabels[e] = value;
      }
      return this;
    }

    if (!isUndefined(name) && !this._isMultigraph) {
      throw new Error("Cannot set a named edge when isMultigraph = false");
    }

    // It didn't exist, so we need to create it.
    // First ensure the nodes exist.
    this.setNode(v);
    this.setNode(w);

    this._edgeLabels[e] = valueSpecified
      ? value
      : this._defaultEdgeLabelFn(v, w, name);

    const edgeObj = edgeArgsToObj(this._isDirected, v, w, name);
    // Ensure we add undirected edges in a consistent way.
    v = edgeObj.v;
    w = edgeObj.w;

    Object.freeze(edgeObj);
    this._edgeObjs[e] = edgeObj;
    incrementOrInitEntry(this._preds[w], v);
    incrementOrInitEntry(this._sucs[v], w);
    this._in[w][e] = edgeObj;
    this._out[v][e] = edgeObj;
    this._edgeCount++;
    return this;
  }

  edge(v, w, name) {
    const e =
      arguments.length === 1
        ? edgeObjToId(this._isDirected, arguments[0])
        : edgeArgsToId(this._isDirected, v, w, name);
    return this._edgeLabels[e];
  }

  hasEdge(v, w, name) {
    const e =
      arguments.length === 1
        ? edgeObjToId(this._isDirected, arguments[0])
        : edgeArgsToId(this._isDirected, v, w, name);
    return has(this._edgeLabels, e);
  }

  removeEdge(v, w, name) {
    const e =
      arguments.length === 1
        ? edgeObjToId(this._isDirected, arguments[0])
        : edgeArgsToId(this._isDirected, v, w, name);
    const edge = this._edgeObjs[e];
    if (edge) {
      v = edge.v;
      w = edge.w;
      delete this._edgeLabels[e];
      delete this._edgeObjs[e];
      decrementOrRemoveEntry(this._preds[w], v);
      decrementOrRemoveEntry(this._sucs[v], w);
      delete this._in[w][e];
      delete this._out[v][e];
      this._edgeCount--;
    }
    return this;
  }

  inEdges(v, u) {
    const inV = this._in[v];
    if (inV) {
      const edges = values(inV);
      if (!u) {
        return edges;
      }
      return filter(edges, (edge) => edge.v === u);
    }
  }

  outEdges(v, w) {
    const outV = this._out[v];
    if (outV) {
      const edges = values(outV);
      if (!w) {
        return edges;
      }
      return filter(edges, (edge) => edge.w === w);
    }
  }

  nodeEdges(v, w) {
    const inEdges = this.inEdges(v, w);
    if (inEdges) {
      return inEdges.concat(this.outEdges(v, w));
    }
  }
}

/* Number of nodes in the graph. Should only be changed by the implementation. */
Graph.prototype._nodeCount = 0;

/* Number of edges in the graph. Should only be changed by the implementation. */
Graph.prototype._edgeCount = 0;

function incrementOrInitEntry(map, k) {
  if (map[k]) {
    map[k]++;
  } else {
    map[k] = 1;
  }
}

function decrementOrRemoveEntry(map, k) {
  if (!--map[k]) {
    delete map[k];
  }
}

function edgeArgsToId(isDirected, v_, w_, name) {
  let v = `${v_}`;
  let w = `${w_}`;
  if (!isDirected && v > w) {
    const tmp = v;
    v = w;
    w = tmp;
  }
  return (
    v +
    EDGE_KEY_DELIM +
    w +
    EDGE_KEY_DELIM +
    (isUndefined(name) ? DEFAULT_EDGE_NAME : name)
  );
}

function edgeArgsToObj(isDirected, v_, w_, name) {
  let v = `${v_}`;
  let w = `${w_}`;
  if (!isDirected && v > w) {
    const tmp = v;
    v = w;
    w = tmp;
  }
  const edgeObj = { v, w };
  if (name) {
    edgeObj.name = name;
  }
  return edgeObj;
}

function edgeObjToId(isDirected, { v, w, name }) {
  return edgeArgsToId(isDirected, v, w, name);
}
export { Graph };
