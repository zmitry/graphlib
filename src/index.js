export * from "./graph";
export { writeOne as writeDot } from "./dot";
export { write as writeJSON, read as readJSON } from "./json";
