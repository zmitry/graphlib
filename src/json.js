import { Graph } from "./graph";

export function write(g) {
  var json = {
    options: {
      directed: g.isDirected(),
      multigraph: g.isMultigraph(),
      compound: g.isCompound()
    },
    nodes: writeNodes(g),
    edges: writeEdges(g)
  };
  if (g.graph() !== undefined) {
    json.value = _.clone(g.graph());
  }
  return json;
}

function writeNodes(g) {
  return g.nodes().map(function(v) {
    var nodeValue = g.node(v);
    var parent = g.parent(v);
    var node = { v: v };
    if (nodeValue !== undefined) {
      node.value = nodeValue;
    }
    if (!parent !== undefined) {
      node.parent = parent;
    }
    return node;
  });
}

function writeEdges(g) {
  return (g.edges() || []).map(function(e) {
    var edgeValue = g.edge(e);
    var edge = { v: e.v, w: e.w };
    if (e.name !== undefined) {
      edge.name = e.name;
    }
    if (edgeValue !== undefined) {
      edge.value = edgeValue;
    }
    return edge;
  });
}

export function read(json) {
  var g = new Graph(json.options).setGraph(json.value);
  for (let entry of json.nodes) {
    g.setNode(entry.v, entry.value);
    if (entry.parent) {
      g.setParent(entry.v, entry.parent);
    }
  }
  for (let entry of json.edges) {
    g.setEdge({ v: entry.v, w: entry.w, name: entry.name }, entry.value);
  }
  return g;
}
