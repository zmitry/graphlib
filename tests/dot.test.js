import { Graph, writeDot } from "../src";

describe("write", function() {
  it("can write subgraphs with attributes", function() {
    var g = new Graph({ compound: true, multigraph: true });
    g.setParent("n1", "root");
    g.setNode("root", { foo: "bar" });
    g.setEdge("n1", "n2", { foo: "baz" }, "another");
    var str = writeDot(g);
    expect(str).toMatchSnapshot();
  });

  it("can write subgraphs with attributes and intendation", function() {
    var g = new Graph({ compound: true, multigraph: true });
    g.setParent("n1", "root");
    g.setNode("root", { foo: "bar" });
    g.setEdge("n1", "n2", { foo: "baz" }, "another");
    var str = writeDot(g, "   ");
    expect(str).toMatchSnapshot();
  });
});
