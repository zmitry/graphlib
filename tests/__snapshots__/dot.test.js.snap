// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`write can write subgraphs with attributes 1`] = `
"digraph {
subgraph root {
foo=bar;
n1
}
n2
n1 -> n2 [foo=baz]
}
"
`;

exports[`write can write subgraphs with attributes and intendation 1`] = `
"digraph {
   subgraph root {
      foo=bar;
      n1
   }
   n2
   n1 -> n2 [foo=baz]
}
"
`;
